### ISSUE TITLE
Unexpected negative probabilities from negative utilities in `chooses` statement

### ISSUE THREAD

**Issue by iancero** [kach/memo#93]
Hi Kartik,  

Thanks again for the great work on `memo`. While using it, I noticed some unexpected behavior when passing negative utilities into the `chooses` statement. Specifically, the resulting probabilities sometimes include negative values, which suggests something unusual may be happening during normalization.  

**Here’s a minimal reproducible example:**  

```python
import jax
import jax.numpy as jnp
from memo import memo
from enum import IntEnum

class Action(IntEnum):
    A = 0
    B = 1
    C = 2

@jax.jit
def some_neg_utilities(a):
    return jnp.array([1, -2, -3])[a]

@memo
def model[a: Action]():
    agent: chooses(a in Action, wpp=some_neg_utilities(a))
    return Pr[agent.a == a]

print(model())  # Output: [-0.25  0.5   0.75]
```

**What I Found**:
1. When utilities are a mix of positive and negative values, at least some of resulting probabilities will be negative.  
2. When all utilities are negative, the probabilities are all positive, but ordinally inconsistent with what you'd expect. For example, using utilities `[-1, -2, -3]` produces probabilities `[0.16, 0.33, 0.50]`, which suggests an agent is actually seeking what it believes are the worst outcomes.

My guess is that this could be happening from simply dividing all array values by their sum, during normalization (which is totally reasonable of memo to do). That is, when some of the utility values are negative, then at least one of the utility values will have a different sign than the sum of utilities, leading to at least one negative probability value. In contrast, when all values are negative, this means we're dividing each negative value by a negative sum, consistently returning positive values - but with the most negative utilities being associated now with the most positive probabilities.

**Proposal**  
I've heard of at least some theoretical uses for negative probabilities, but I think they are a fairly uncommon way of thinking about probability overall. 
- For most users then, it might be most helpful for **memo to throw an error when `Pr()` produces any negative values**. 
- On the other hand, if we do generally want to preserve the option for producing negative probabilities, could `Pr()` still throw a silence-able warning? Thoughts?
- One remaining problem here is that even if memo errors for negative probabilities, that **wouldn't solve the "all negative utilities produces probabilities that are opposite of an agent's preferences" problem**. I honestly don't know how to handle that issue, but I'm happy to discuss more too.

Thanks again for your time and work on `memo`. I’m happy to help debug further if needed!  

Best,  
Ian

**Comment by kach** [comment]
Hi Ian,

Thanks for the report! Your diagnosis is exactly right: memo evaluates the `wpp=` expression for each possible choice, and then normalizes by dividing by the sum. When the specified probabilities are negative, weird things happen.

An important point is that `wpp=e` should be read as "with probability proportional to _e_", and _not_ as "with utility _e_." While utilities can be negative, probabilities cannot be negative. If _e_ evaluates to a negative number, that is symptomatic of a mistake. As you suggested, it would be really nice to be able to raise an error message if that ever happens. But here's the problem: performing that check every time an agent makes a choice would actually slow down memo considerably! So for now, I have been reluctant to add that check, instead putting the burden on memo users to be careful not to specify negative probabilities.

Now, to make a choice with a _utility_ (not a raw probability), we usually write `wpp=exp(Utility(a))`. This makes a choice by "softmax" over the utility, which is by far the most common way in which people translate utilities into probability distributions over choices. There are many theoretical reasons why this really is the "right answer" for translating utilities to probabilities (see Franke, M., & Degen, J. (2023). _The softmax function: Properties, motivation, and interpretation_ for more!). Importantly in this case, notice that `exp(x)` is always positive, so there is no risk of negative probabilities, even if the utilities themselves are negative. So I think your immediate problem would be solved if you adopted the softmax approach.

Does that address your question?

**Comment by iancero** [comment]
Hi Kartik,

Thanks for the quick and thoughtful response (and for your patience on mine here)! That all makes perfect sense—especially the distinction between probabilities and utilities, and the reason why memo doesn’t explicitly check for negative values in wpp for efficiency reasons.

Using wpp=exp(Utility(a)) to enforce non-negative probabilities sounds like the right move. I appreciate the pointer to Franke & Degen (2023) as well—I’ll give it a read!

Thanks again, memo has been really exciting to use!

Best,
Ian
